File: api/parsers/detail_parser.py
"""
Enhanced detail-page parser.

Extracts the full set of metadata from a single movie's detail page
including title, code, duration, maker/publisher, series, directors, tags,
rating, poster, fanart, trailer, actors, magnets, review stats, etc.

No ``time.sleep`` calls are made – the caller controls request pacing.
"""

from __future__ import annotations

import re
import logging
from typing import Optional

from bs4 import BeautifulSoup
from bs4.element import Tag

from api.models import ActorCredit, MovieDetail, MovieLink, MagnetInfo, NO_ACTOR_LISTING_ACTOR_NAME
from api.parsers.common import (
    extract_rate_and_comments,
    extract_movie_link,
    extract_all_movie_links,
)

logger = logging.getLogger(__name__)

# Plain-text values in 演員 span.value when there are no /actors/ links (normalised to NO_ACTOR_LISTING_ACTOR_NAME in DB).
_NO_ACTOR_LISTING_PLACEHOLDERS = frozenset({
    'N/A',
    'n/a',
    '无',
    '無',
    '—',
    '–',
    '-',
    '－',
    '暂无',
    '暂无演员',
})

# Label aliases for locale-independent panel-block matching.
# JavDB supports Traditional Chinese (zh) and English (en).  CF bypass may
# return the English locale when the headless browser lacks a locale cookie.
# Matching is exact (``strong.get_text(strip=True) == label``).
_L_CODE = ('番號:', 'ID:')
_L_DATE = ('日期:', 'Released Date:')
_L_DURATION = ('時長:', 'Duration:')
_L_DIRECTOR = ('導演:', 'Director:')
_L_MAKER = ('片商:', 'Maker:')
_L_PUBLISHER = ('發行商:', 'Publisher:')
_L_SERIES = ('系列:', 'Series:')
_L_RATING = ('評分:', 'Rating:')
_L_TAGS = ('類別:', 'Tags:')
_L_ACTOR = ('演員:', 'Actor(s):')


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------

def _find_panel_block(panel_blocks, label) -> Optional[Tag]:
    """Find the ``<div class="panel-block">`` whose ``<strong>`` text
    exactly matches *label*.

    *label* can be a single string (``"片商:"``) or a tuple of alternatives
    (``("片商:", "Maker:")``).  The first matching block is returned.
    """
    labels = (label,) if isinstance(label, str) else label
    for block in panel_blocks:
        strong = block.find('strong')
        if strong:
            text = strong.get_text(strip=True)
            if text in labels:
                return block
    return None


def _extract_link_from_panel(panel_blocks, label) -> Optional[MovieLink]:
    """Extract a single ``MovieLink`` from a panel-block identified by *label*."""
    block = _find_panel_block(panel_blocks, label)
    if not block:
        return None
    value_span = block.find('span', class_='value')
    if not value_span:
        return None
    a_tag = value_span.find('a')
    return extract_movie_link(a_tag)


def _extract_links_from_panel(panel_blocks, label) -> list:
    """Extract all ``MovieLink`` objects from a panel-block."""
    block = _find_panel_block(panel_blocks, label)
    if not block:
        return []
    value_span = block.find('span', class_='value')
    if not value_span:
        return []
    return extract_all_movie_links(value_span)


def _actor_gender_from_following_marker(a_tag: Tag) -> str:
    """Read ``female`` / ``male`` from the ``<strong class=\"symbol …\">`` after *a_tag*."""
    sib = a_tag.next_sibling
    while sib is not None and not isinstance(sib, Tag):
        sib = sib.next_sibling
    if sib is None or sib.name != 'strong':
        return ''
    classes = sib.get('class') or []
    if isinstance(classes, str):
        classes = classes.split()
    if 'female' in classes:
        return 'female'
    if 'male' in classes:
        return 'male'
    return ''


def _find_actor_panel_by_links(panel_blocks) -> Optional[Tag]:
    """Structural fallback: find a panel-block containing ``/actors/`` links.

    Used when locale-based label matching fails (e.g. CF bypass returning a
    page in simplified Chinese or English).
    """
    for block in panel_blocks:
        value_span = block.find('span', class_='value')
        if not value_span:
            continue
        for a_tag in value_span.find_all('a'):
            href = (a_tag.get('href') or '').strip()
            if '/actors/' in href:
                return block
    return None


def _extract_actors_with_gender(panel_blocks) -> list:
    """Actors from the 演員 panel, in order, with gender from ♀/♂ markers."""
    block = _find_panel_block(panel_blocks, _L_ACTOR)
    if not block:
        block = _find_actor_panel_by_links(panel_blocks)
    if not block:
        return []
    value_span = block.find('span', class_='value')
    if not value_span:
        return []
    actors: list = []
    for a_tag in value_span.find_all('a'):
        if not isinstance(a_tag, Tag):
            continue
        href = (a_tag.get('href') or '').strip()
        if '/actors/' not in href:
            continue
        ml = extract_movie_link(a_tag)
        if not ml:
            continue
        gender = _actor_gender_from_following_marker(a_tag)
        actors.append(ActorCredit(name=ml.name, href=ml.href, gender=gender))
    return actors


def _value_span_has_actor_link(value_span) -> bool:
    if not value_span:
        return False
    for a_tag in value_span.find_all('a'):
        if not isinstance(a_tag, Tag):
            continue
        href = (a_tag.get('href') or '').strip()
        if '/actors/' in href:
            return True
    return False


def _is_no_actor_placeholder_text(text: str) -> bool:
    t = text.strip()
    if not t:
        return False
    if t.casefold() == 'n/a':
        return True
    return t in _NO_ACTOR_LISTING_PLACEHOLDERS


def _detect_no_actor_listing(panel_blocks) -> bool:
    """True when 演員 panel exists, has no actor links, and value text is a known placeholder."""
    block = _find_panel_block(panel_blocks, _L_ACTOR)
    if not block:
        return False
    value_span = block.find('span', class_='value')
    if not value_span:
        return False
    if _value_span_has_actor_link(value_span):
        return False
    raw = value_span.get_text(strip=True)
    return _is_no_actor_placeholder_text(raw)


def _extract_text_from_panel(panel_blocks, label) -> str:
    """Extract plain text from a panel-block."""
    block = _find_panel_block(panel_blocks, label)
    if not block:
        return ''
    value_span = block.find('span', class_='value')
    if not value_span:
        return ''
    return value_span.get_text(strip=True)


def _parse_magnets(soup: BeautifulSoup) -> tuple:
    """Parse magnet links from ``#magnets-content``.

    Returns:
        (magnets_list, parse_success)
    """
    magnets_content = soup.find('div', id='magnets-content')
    if not magnets_content:
        return [], False

    magnets = []
    for item in magnets_content.find_all('div', class_=re.compile(r'item columns is-desktop')):
        magnet_name_div = item.find('div', class_='magnet-name')
        if not magnet_name_div:
            continue

        magnet_a = magnet_name_div.find('a')
        if not magnet_a:
            continue

        magnet_href = magnet_a.get('href', '')
        name_span = magnet_a.find('span', class_='name')
        name = name_span.get_text(strip=True) if name_span else ''

        # Size and file count
        size = ''
        file_count = 0
        meta_span = magnet_a.find('span', class_='meta')
        if meta_span:
            meta_text = meta_span.get_text(strip=True)
            size_match = re.search(r'([\d.]+\s*(?:GB|MB|KB|TB))', meta_text)
            if size_match:
                size = size_match.group(1).replace(' ', '')
            fc_match = re.search(r'(\d+)\s*(?:個文件|files?)', meta_text)
            if fc_match:
                file_count = int(fc_match.group(1))

        # Timestamp
        timestamp = ''
        time_span = item.find('span', class_='time')
        if time_span:
            timestamp = time_span.get_text(strip=True)

        # Tags
        tags = []
        tags_div = magnet_a.find('div', class_='tags')
        if tags_div and isinstance(tags_div, Tag):
            for span in tags_div.find_all('span', class_='tag'):
                if isinstance(span, Tag):
                    tags.append(span.get_text(strip=True))

        magnets.append(MagnetInfo(
            href=magnet_href,
            name=name,
            tags=tags,
            size=size,
            file_count=file_count,
            timestamp=timestamp,
        ))

    return magnets, True


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def parse_detail_page(html_content: str) -> MovieDetail:
    """Parse a movie detail page and return a fully-populated *MovieDetail*.

    Fields that are not present on the page are left at their default values
    (empty string / empty list / None / 0).
    """
    soup = BeautifulSoup(html_content, 'html.parser')
    detail = MovieDetail()

    # --- Title (from <h2> containing <strong class="current-title">) ---
    title_strong = soup.find('strong', class_='current-title')
    if title_strong:
        detail.title = title_strong.get_text(strip=True)

    # --- Metadata panel ---
    video_meta_panel = soup.find('div', class_='video-meta-panel')
    panel_blocks = []
    if video_meta_panel:
        panel_blocks = video_meta_panel.find_all('div', class_='panel-block')

    # --- Video code + prefix link ---
    first_block = _find_panel_block(panel_blocks, _L_CODE)
    if first_block:
        value_span = first_block.find('span', class_='value')
        if value_span:
            # The full video code text (e.g. "VDD-201")
            detail.video_code = value_span.get_text(strip=True)
            # The prefix link (e.g. /video_codes/VDD)
            prefix_a = value_span.find('a')
            if prefix_a:
                detail.code_prefix_link = prefix_a.get('href', '')

    # --- Release date ---
    detail.release_date = _extract_text_from_panel(panel_blocks, _L_DATE)

    # --- Duration ---
    detail.duration = _extract_text_from_panel(panel_blocks, _L_DURATION)

    # --- Directors ---
    detail.directors = _extract_links_from_panel(panel_blocks, _L_DIRECTOR)

    # --- Maker (片商) ---
    detail.maker = _extract_link_from_panel(panel_blocks, _L_MAKER)

    # --- Publisher (發行商) – may not be present on all pages ---
    detail.publisher = _extract_link_from_panel(panel_blocks, _L_PUBLISHER)

    # --- Series ---
    detail.series = _extract_link_from_panel(panel_blocks, _L_SERIES)

    # --- Rating & comment count ---
    rating_block = _find_panel_block(panel_blocks, _L_RATING)
    if rating_block:
        value_span = rating_block.find('span', class_='value')
        if value_span:
            score_text = value_span.get_text(strip=True)
            detail.rate, detail.comment_count = extract_rate_and_comments(score_text)

    # --- Tags (類別) ---
    detail.tags = _extract_links_from_panel(panel_blocks, _L_TAGS)

    # --- Actors (演員) — include gender from adjacent symbol markers ---
    detail.actors = _extract_actors_with_gender(panel_blocks)
    if not detail.actors:
        detail.no_actor_listing = _detect_no_actor_listing(panel_blocks)
        if detail.no_actor_listing:
            logger.debug(
                'Detail: no actor links; placeholder 演員 text → %r',
                NO_ACTOR_LISTING_ACTOR_NAME,
            )

    # --- Poster URL (cover image) ---
    if video_meta_panel:
        cover_col = video_meta_panel.find('div', class_='column-video-cover')
        if cover_col:
            cover_img = cover_col.find('img', class_='video-cover')
            if cover_img:
                detail.poster_url = cover_img.get('src', '')

    # --- Fanart URLs (sample images) ---
    tile_images = soup.find('div', class_='tile-images preview-images')
    if tile_images:
        for tile in tile_images.find_all('a', class_='tile-item'):
            href = tile.get('href', '')
            if href:
                detail.fanart_urls.append(href)

    # --- Trailer URL ---
    # The preview video container links to the video section
    preview_container = soup.find('a', class_='preview-video-container')
    if preview_container:
        # The actual video element
        video_el = soup.find('video', id='preview-video')
        if video_el:
            src = video_el.get('src', '')
            if src and not src.startswith('blob:'):
                detail.trailer_url = src
            else:
                # Try data-src or source element
                source = video_el.find('source')
                if source:
                    detail.trailer_url = source.get('src', '') or None
        # If no direct src, note the preview container href
        if not detail.trailer_url:
            container_href = preview_container.get('href', '')
            if container_href and '#preview-video' in container_href:
                # The trailer is loaded dynamically; store the base URL
                detail.trailer_url = container_href

    # --- Review count (from tabs "短評(N)" / "Reviews(N)") ---
    review_tab = soup.find('a', class_='review-tab')
    if review_tab:
        tab_text = review_tab.get_text(strip=True)
        review_match = re.search(r'(?:短評|Reviews)\((\d+)\)', tab_text)
        if review_match:
            detail.review_count = int(review_match.group(1))

    # --- Want to watch / watched counts ---
    for block in panel_blocks:
        span = block.find('span', class_='is-size-7')
        if span:
            text = span.get_text(strip=True)
            want_match = re.search(r'(\d+)\s*(?:人想看|want to watch)', text)
            if want_match:
                detail.want_count = int(want_match.group(1))
            watched_match = re.search(r'(\d+)\s*(?:人看過|have seen)', text)
            if watched_match:
                detail.watched_count = int(watched_match.group(1))

    # --- Magnets ---
    magnets, parse_success = _parse_magnets(soup)
    detail.magnets = magnets
    detail.parse_success = parse_success

    logger.debug(
        'Parsed detail: code=%s, title=%s, actors=%d, magnets=%d',
        detail.video_code,
        detail.title[:40] if detail.title else '',
        len(detail.actors),
        len(detail.magnets),
    )

    return detail

File: rust_core/src/scraper/detail_parser.rs
use log::debug;
use once_cell::sync::Lazy;
use regex::Regex;
use scraper::{ElementRef, Html, Selector};

use crate::models::{MagnetInfo, MovieDetail, MovieLink};
use crate::scraper::common::{
    extract_all_movie_links, extract_movie_link, extract_rate_and_comments, get_text_content,
};

static SEL_CURRENT_TITLE: Lazy<Selector> =
    Lazy::new(|| Selector::parse("strong.current-title").unwrap());
static SEL_VIDEO_META_PANEL: Lazy<Selector> =
    Lazy::new(|| Selector::parse("div.video-meta-panel").unwrap());
static SEL_PANEL_BLOCK: Lazy<Selector> =
    Lazy::new(|| Selector::parse("div.panel-block").unwrap());
static SEL_STRONG: Lazy<Selector> = Lazy::new(|| Selector::parse("strong").unwrap());
static SEL_VALUE: Lazy<Selector> = Lazy::new(|| Selector::parse("span.value").unwrap());
static SEL_A: Lazy<Selector> = Lazy::new(|| Selector::parse("a").unwrap());
static SEL_MAGNETS_CONTENT: Lazy<Selector> =
    Lazy::new(|| Selector::parse("div#magnets-content").unwrap());
static SEL_MAGNET_NAME: Lazy<Selector> =
    Lazy::new(|| Selector::parse("div.magnet-name").unwrap());
static SEL_NAME_SPAN: Lazy<Selector> = Lazy::new(|| Selector::parse("span.name").unwrap());
static SEL_META_SPAN: Lazy<Selector> = Lazy::new(|| Selector::parse("span.meta").unwrap());
static SEL_TIME_SPAN: Lazy<Selector> = Lazy::new(|| Selector::parse("span.time").unwrap());
static SEL_TAGS_DIV: Lazy<Selector> = Lazy::new(|| Selector::parse("div.tags").unwrap());
static SEL_TAG_SPAN: Lazy<Selector> = Lazy::new(|| Selector::parse("span.tag").unwrap());
static SEL_COVER_COL: Lazy<Selector> =
    Lazy::new(|| Selector::parse("div.column-video-cover").unwrap());
static SEL_COVER_IMG: Lazy<Selector> =
    Lazy::new(|| Selector::parse("img.video-cover").unwrap());
static SEL_TILE_IMAGES: Lazy<Selector> =
    Lazy::new(|| Selector::parse("div.tile-images.preview-images").unwrap());
static SEL_TILE_ITEM: Lazy<Selector> = Lazy::new(|| Selector::parse("a.tile-item").unwrap());
static SEL_PREVIEW_CONTAINER: Lazy<Selector> =
    Lazy::new(|| Selector::parse("a.preview-video-container").unwrap());
static SEL_PREVIEW_VIDEO: Lazy<Selector> =
    Lazy::new(|| Selector::parse("video#preview-video").unwrap());
static SEL_SOURCE: Lazy<Selector> = Lazy::new(|| Selector::parse("source").unwrap());
static SEL_REVIEW_TAB: Lazy<Selector> =
    Lazy::new(|| Selector::parse("a.review-tab").unwrap());
static SEL_SIZE7: Lazy<Selector> = Lazy::new(|| Selector::parse("span.is-size-7").unwrap());

static MAGNET_ITEM_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"item columns is-desktop").unwrap());
static SIZE_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"([\d.]+(?:GB|MB|KB|TB))").unwrap());
static REVIEW_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"短評\((\d+)\)").unwrap());
static WANT_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"(\d+)人想看").unwrap());
static WATCHED_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"(\d+)人看過").unwrap());

fn find_panel_block<'a>(
    panel_blocks: &[ElementRef<'a>],
    label: &str,
) -> Option<ElementRef<'a>> {
    panel_blocks.iter().find(|block| {
        block
            .select(&SEL_STRONG)
            .next()
            .map_or(false, |strong| get_text_content(&strong).contains(label))
    }).copied()
}

fn extract_link_from_panel(panel_blocks: &[ElementRef], label: &str) -> Option<MovieLink> {
    let block = find_panel_block(panel_blocks, label)?;
    let value_span = block.select(&SEL_VALUE).next()?;
    let a_tag = value_span.select(&SEL_A).next()?;
    extract_movie_link(&a_tag)
}

fn extract_links_from_panel(panel_blocks: &[ElementRef], label: &str) -> Vec<MovieLink> {
    let block = match find_panel_block(panel_blocks, label) {
        Some(b) => b,
        None => return Vec::new(),
    };
    let value_span = match block.select(&SEL_VALUE).next() {
        Some(v) => v,
        None => return Vec::new(),
    };
    extract_all_movie_links(&value_span)
}

fn extract_text_from_panel(panel_blocks: &[ElementRef], label: &str) -> String {
    let block = match find_panel_block(panel_blocks, label) {
        Some(b) => b,
        None => return String::new(),
    };
    block
        .select(&SEL_VALUE)
        .next()
        .map_or(String::new(), |v| get_text_content(&v).trim().to_string())
}

fn parse_magnets(document: &Html) -> (Vec<MagnetInfo>, bool) {
    let magnets_content = match document.select(&SEL_MAGNETS_CONTENT).next() {
        Some(mc) => mc,
        None => return (Vec::new(), false),
    };

    let mut magnets = Vec::new();

    // Find all magnet items by class pattern
    for item in magnets_content
        .descendants()
        .filter_map(|n| ElementRef::wrap(n))
        .filter(|el| {
            el.value().name() == "div"
                && el
                    .value()
                    .attr("class")
                    .map_or(false, |c| MAGNET_ITEM_RE.is_match(c))
        })
    {
        let magnet_name_div = match item.select(&SEL_MAGNET_NAME).next() {
            Some(d) => d,
            None => continue,
        };

        let magnet_a = match magnet_name_div.select(&SEL_A).next() {
            Some(a) => a,
            None => continue,
        };

        let magnet_href = magnet_a.value().attr("href").unwrap_or("").to_string();
        let name = magnet_a
            .select(&SEL_NAME_SPAN)
            .next()
            .map_or(String::new(), |s| get_text_content(&s).trim().to_string());

        // Size
        let size = magnet_a
            .select(&SEL_META_SPAN)
            .next()
            .and_then(|meta| {
                let meta_text = get_text_content(&meta).trim().to_string();
                SIZE_RE
                    .captures(&meta_text)
                    .and_then(|c| c.get(1))
                    .map(|m| m.as_str().to_string())
            })
            .unwrap_or_default();

        // Timestamp
        let timestamp = item
            .select(&SEL_TIME_SPAN)
            .next()
            .map_or(String::new(), |t| get_text_content(&t).trim().to_string());

        // Tags
        let mut tags = Vec::new();
        if let Some(tags_div) = magnet_a.select(&SEL_TAGS_DIV).next() {
            for span in tags_div.select(&SEL_TAG_SPAN) {
                let tag_text = get_text_content(&span).trim().to_string();
                if !tag_text.is_empty() {
                    tags.push(tag_text);
                }
            }
        }

        magnets.push(MagnetInfo {
            href: magnet_href,
            name,
            tags,
            size,
            timestamp,
        });
    }

    (magnets, true)
}

pub fn parse_detail_page(html_content: &str) -> MovieDetail {
    let document = Html::parse_document(html_content);
    let mut detail = MovieDetail::default();

    // Title
    if let Some(title_strong) = document.select(&SEL_CURRENT_TITLE).next() {
        detail.title = get_text_content(&title_strong).trim().to_string();
    }

    // Metadata panel
    let video_meta_panel = document.select(&SEL_VIDEO_META_PANEL).next();
    let panel_blocks: Vec<ElementRef> = video_meta_panel
        .map(|p| p.select(&SEL_PANEL_BLOCK).collect())
        .unwrap_or_default();

    // Video code + prefix link
    if let Some(first_block) = find_panel_block(&panel_blocks, "番號:") {
        if let Some(value_span) = first_block.select(&SEL_VALUE).next() {
            detail.video_code = get_text_content(&value_span).trim().to_string();
            if let Some(prefix_a) = value_span.select(&SEL_A).next() {
                detail.code_prefix_link =
                    prefix_a.value().attr("href").unwrap_or("").to_string();
            }
        }
    }

    // Release date, Duration
    detail.release_date = extract_text_from_panel(&panel_blocks, "日期:");
    detail.duration = extract_text_from_panel(&panel_blocks, "時長:");

    // Directors, Maker, Publisher, Series
    detail.directors = extract_links_from_panel(&panel_blocks, "導演:");
    detail.maker = extract_link_from_panel(&panel_blocks, "片商:");
    detail.publisher = extract_link_from_panel(&panel_blocks, "發行商:");
    detail.series = extract_link_from_panel(&panel_blocks, "系列:");

    // Rating & comment count
    if let Some(rating_block) = find_panel_block(&panel_blocks, "評分:") {
        if let Some(value_span) = rating_block.select(&SEL_VALUE).next() {
            let score_text = get_text_content(&value_span).trim().to_string();
            let (r, c) = extract_rate_and_comments(&score_text);
            detail.rate = r;
            detail.comment_count = c;
        }
    }

    // Tags, Actors
    detail.tags = extract_links_from_panel(&panel_blocks, "類別:");
    detail.actors = extract_links_from_panel(&panel_blocks, "演員:");

    // Poster URL
    if let Some(vmp) = video_meta_panel {
        if let Some(cover_col) = vmp.select(&SEL_COVER_COL).next() {
            if let Some(cover_img) = cover_col.select(&SEL_COVER_IMG).next() {
                detail.poster_url = cover_img
                    .value()
                    .attr("src")
                    .unwrap_or("")
                    .to_string();
            }
        }
    }

    // Fanart URLs
    if let Some(tile_images) = document.select(&SEL_TILE_IMAGES).next() {
        for tile in tile_images.select(&SEL_TILE_ITEM) {
            let href = tile.value().attr("href").unwrap_or("").to_string();
            if !href.is_empty() {
                detail.fanart_urls.push(href);
            }
        }
    }

    // Trailer URL
    if let Some(preview_container) = document.select(&SEL_PREVIEW_CONTAINER).next() {
        if let Some(video_el) = document.select(&SEL_PREVIEW_VIDEO).next() {
            let src = video_el.value().attr("src").unwrap_or("").to_string();
            if !src.is_empty() && !src.starts_with("blob:") {
                detail.trailer_url = Some(src);
            } else if let Some(source) = video_el.select(&SEL_SOURCE).next() {
                let source_src = source.value().attr("src").unwrap_or("").to_string();
                if !source_src.is_empty() {
                    detail.trailer_url = Some(source_src);
                }
            }
        }
        if detail.trailer_url.is_none() {
            let container_href = preview_container.value().attr("href").unwrap_or("");
            if container_href.contains("#preview-video") {
                detail.trailer_url = Some(container_href.to_string());
            }
        }
    }

    // Review count
    if let Some(review_tab) = document.select(&SEL_REVIEW_TAB).next() {
        let tab_text = get_text_content(&review_tab).trim().to_string();
        if let Some(caps) = REVIEW_RE.captures(&tab_text) {
            if let Ok(count) = caps[1].parse::<i32>() {
                detail.review_count = count;
            }
        }
    }

    // Want/Watched counts
    for block in &panel_blocks {
        if let Some(span) = block.select(&SEL_SIZE7).next() {
            let text = get_text_content(&span).trim().to_string();
            if let Some(caps) = WANT_RE.captures(&text) {
                if let Ok(count) = caps[1].parse::<i32>() {
                    detail.want_count = count;
                }
            }
            if let Some(caps) = WATCHED_RE.captures(&text) {
                if let Ok(count) = caps[1].parse::<i32>() {
                    detail.watched_count = count;
                }
            }
        }
    }

    // Magnets
    let (magnets, parse_success) = parse_magnets(&document);
    detail.magnets = magnets;
    detail.parse_success = parse_success;

    debug!(
        "Parsed detail: code={}, title={}, actors={}, magnets={}",
        detail.video_code,
        if detail.title.len() > 40 {
            &detail.title[..40]
        } else {
            &detail.title
        },
        detail.actors.len(),
        detail.magnets.len(),
    );

    detail
}
